### PR TITLE
Parallel backgroundjob execution

### PR DESCRIPTION
**Pull request by icewind1991**
Removes the global lock from cron.php and replaces it with a more fine grained locking system to allow concurrent job execution.
- A global lock for the joblist is acquired each time the job executor gets the next job from the list
  this ensures that multiple executors running at the same time will get different jobs to run
- A lock is acquired on a job when it's starts executing it to prevent multiple process from touching the same job in cases of long running jobs
- A 5 minute minimum delay is added before any job is considered for execution again after it's run, this prevents multiple executers running the same job after each other unnecessarily
- An occ command (`occ background:run`) has been added as an additional way to trigger the execution of background jobs, using this an admin can create more sophisticated system for running jobs.

I've done some tested (create artificial long running jobs, start cron.php a few time and look which cron.php instance runs which job) but ideally we'll figure out some way to get automated testing of parallel execution

cc @DeepDiver1975 @PVince81 @MorrisJobke

### CODE REVIEW

File: lib/private/backgroundjob/executer.php [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+		} else {
+			return false;
+		}
+	}
+
+	public function getNextJob() {
+		$timeWaited = 0;
+		while ($timeWaited < self::LOCK_TIMEOUT) {
+			try {
+				$this->lockingProvider->acquireLock('joblist', ILockingProvider::LOCK_EXCLUSIVE);
```
Comment (DeepDiver1975) [inline lib/private/backgroundjob/executer.php]: Why use an external locking provider?

SQL comes with all necessary capabilities to implement this in one source - not in two

I'm sorry - but did you actually read my proposal?
https://github.com/owncloud/core/issues/23544#issuecomment-203862962

**Comment by BernhardPosselt** [comment]
Could this cause concurrency issues?

**Comment by icewind1991** [comment]
The locking that is in place should prevent any concurrency issues

**Comment by MorrisJobke** [comment]
> A 5 minute minimum delay is added before any job is considered for execution again after it's run, this prevents multiple executers running the same job after each other unnecessarily

I don't think, that we want this.

**Comment by nickvergessen** [comment]
> > A 5 minute minimum delay is added before any job is considered for execution again after it's run, this prevents multiple executers running the same job after each other unnecessarily
>
> I don't think, that we want this.

But we want something like this. @schiesbn created an infinite loop otherwise with a background job, that spawns a child of the same class

**Comment by MorrisJobke** [comment]
> But we want something like this. @schiesbn created an infinite loop otherwise with a background job, that spawns a child of the same class

Then the looping is incorrect. What if an job needs to be executed every 2 minutes.

**Comment by nickvergessen** [comment]
As per our docs cron should be called all 15mins, that needs adjusting, anyway thou

**Comment by DeepDiver1975** [comment]
Forget about cron. We need to have many executors running in parallel - actually with high throughput of jobs.
This requires executor process living permanent listening (yes Polling) the DB Queue.

This is why the design of laravel Queue listeners was proposed.

Please implement accordingly or bring valid arguments against this architectural proposal.

**Pull request closed without merging**